src/SExpr.purs

```diff
@@ -472,6 +496,33 @@ clausesStateBwd κ0 ks = case κ0 × ks of
       kss = defined (popConstrFwd (defined (dataTypeFor (definitely' (ctrFor p)))) ks)
    ContElim _ × _ -> error (shapeMismatch unit)
 
+desugComment :: ∀ m. MonadError Error m => Maybe Comment -> m (Maybe E.Comment)
+desugComment Nothing = pure Nothing
+desugComment (Just c) = Just <$> commentFwd c
+
+desugCommentBwd :: Maybe E.Comment -> Maybe Comment -> Maybe Comment
+desugCommentBwd Nothing Nothing = Nothing
+desugCommentBwd (Just ec) (Just c) = Just (commentBwd ec c)
+desugCommentBwd _ _ = error "desugCommentBwd mismatch"
+
+commentFwd :: ∀ m. MonadError Error m => Comment -> m E.Comment
+commentFwd (Cons s l) = Cons <$> commentElemFwd s <*> commentFwd l
+commentFwd Nil = pure Nil
+
+commentElemFwd :: ∀ m. MonadError Error m => CommentElem -> m E.CommentElem
+commentElemFwd (Literal s) = pure $ E.Literal s
+commentElemFwd (CExpr e) = E.CExpr <$> exprFwd e
+
+commentBwd :: E.Comment -> Comment -> Comment
+commentBwd (Cons c l) (Cons c' l') = Cons (commentElemBwd c c') (commentBwd l l')
+commentBwd Nil Nil = Nil
+commentBwd _ _ = error "commentBwd mismatch"
+
+commentElemBwd :: E.CommentElem -> CommentElem -> CommentElem
+commentElemBwd (E.Literal _) (Literal s') = Literal s'
+commentElemBwd (E.CExpr e) (CExpr e') = CExpr (exprBwd e e')
+commentElemBwd _ _ = error "commentElemBwd mismatch"
```

Comment:
I don't think we do this entirely consistently in `evalBwd` but I think the convention should probably be `error (shapeMismatch unit)`